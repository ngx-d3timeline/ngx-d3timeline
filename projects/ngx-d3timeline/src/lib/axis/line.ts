import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
