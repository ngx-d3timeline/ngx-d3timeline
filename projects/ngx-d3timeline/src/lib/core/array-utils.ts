export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}
