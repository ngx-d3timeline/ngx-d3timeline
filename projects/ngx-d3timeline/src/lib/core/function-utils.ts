export function partialApply<TInput, TAppliedArgs extends any[], TOut>(
  func: (input: TInput, ...appliedArgs: TAppliedArgs) => TOut
): (...appliedArgs: TAppliedArgs) => (input: TInput) => TOut {
  return (...appliedArgs: TAppliedArgs) => (input: TInput) =>
    func(input, ...appliedArgs);
}

export function pipe<T>(
  ...args: ((...args: any[]) => T)[]
): (...t: any[]) => T {
  return x => args.reduce((acc, curr) => curr(acc), x);
}

export function clampZero(x: number) {
  return Math.max(x, 0);
}
