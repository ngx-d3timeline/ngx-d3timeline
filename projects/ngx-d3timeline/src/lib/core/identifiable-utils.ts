import { Identifiable, identifier } from './types';

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
