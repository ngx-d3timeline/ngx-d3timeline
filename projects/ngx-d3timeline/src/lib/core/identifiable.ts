export type Identifier = string | number;

export interface Identifiable {
  id: Identifier;
}

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: Identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
