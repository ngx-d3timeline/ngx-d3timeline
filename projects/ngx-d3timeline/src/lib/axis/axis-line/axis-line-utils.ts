import { Scale } from '../../scale-types';
import { Orientation } from '../../orientation';
import { TimelineView } from '../../view/timeline-view';
import { Line } from '../line';

export function getAxisLine<TScale extends Scale>(
  scale: TScale,
  orientation: Orientation,
  timelineView: TimelineView
): Line {
  const axisLine: Line = {
    x1: timelineView.left,
    x2: timelineView.left,
    y1: timelineView.top,
    y2: timelineView.top
  };
  const rangeLimit = scale.range()[1];

  return orientation === Orientation.Vertical
    ? { ...axisLine, y2: rangeLimit }
    : { ...axisLine, x2: rangeLimit };
}
