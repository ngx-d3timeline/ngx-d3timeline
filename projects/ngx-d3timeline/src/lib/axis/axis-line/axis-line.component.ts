import { Component, Input } from '@angular/core';
import { Line } from '../line';
@Component({
  selector: '[ngx-d3timeline-axis-line]',
  template: `
    <svg:line
      class="axis-line"
      [attr.x1]="axisLine.x1"
      [attr.x2]="axisLine.x2"
      [attr.y1]="axisLine.y1"
      [attr.y2]="axisLine.y2"
    ></svg:line>
  `,
  styles: [
    `
      .axis-line {
        stroke: #000;
      }
    `
  ]
})
export class AxisLineComponent {
  @Input() axisLine: Line;
}
