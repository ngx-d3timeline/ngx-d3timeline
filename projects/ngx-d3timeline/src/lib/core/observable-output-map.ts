import { Observable, Subject } from 'rxjs';
import { EventEmitter } from '@angular/core';
import { takeUntil } from 'rxjs/operators';

export interface ObservableOutputMap<T> {
  observable$: Observable<T>;
  output: EventEmitter<T>;
}

export function setupObservableToOutputMappings(
  observableToOutputMap: ObservableOutputMap<any>[],
  destroy$: Subject<boolean>
) {
  observableToOutputMap.forEach(map =>
    setupObservableToOutputMapping(map, destroy$)
  );
}

function setupObservableToOutputMapping<T>(
  observableOutputMap: ObservableOutputMap<T>,
  destroy$: Subject<boolean>
) {
  const { observable$, output } = observableOutputMap;
  observable$.pipe(takeUntil(destroy$)).subscribe(value => output.emit(value));
}
