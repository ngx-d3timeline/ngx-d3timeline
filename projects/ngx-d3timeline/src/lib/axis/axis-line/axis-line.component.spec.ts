import { Line } from '../line';
import { TestBed, ComponentFixture } from '@angular/core/testing';
import { AxisLineComponent } from './axis-line.component';
import { Component, Input } from '@angular/core';

describe('AxisLineComponent', () => {
  let fixture: ComponentFixture<AxisLineComponent>;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [AxisLineComponent]
    });

    fixture = TestBed.createComponent(AxisLineComponent);
  });

  it('should render correctly', () => {
    const axisLine: Line = {
      x1: 0,
      x2: 50,
      y1: 0,
      y2: 50
    };
    fixture.componentInstance.axisLine = axisLine;
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
