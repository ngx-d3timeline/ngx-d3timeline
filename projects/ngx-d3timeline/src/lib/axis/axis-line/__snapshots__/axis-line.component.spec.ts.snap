// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxisLineComponent should render correctly 1`] = `
<div
  id="root0"
  ngx-d3timeline-axis-line=""
>
  <line
    class="axis-line"
    x1="0"
    x2="50"
    y1="0"
    y2="50"
  />
</div>
`;
